packages/terminal/src/browser/default-terminal-client.ts

```diff
@@ -0,0 +1,222 @@
+
+import { injectable, inject, postConstruct, named } from 'inversify';
+import { ShellTerminalServerProxy } from '../common/shell-terminal-protocol';
+import { IBaseTerminalServer } from '../common/base-terminal-protocol';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { MessageConnection } from 'vscode-jsonrpc';
+import { WebSocketConnectionProvider } from '@theia/core/lib/browser';
+import { terminalsPath } from '../common/terminal-protocol';
+import { TerminalWatcher } from '../common/terminal-watcher';
+import { ILogger, Disposable, DisposableCollection } from '@theia/core';
+import { TerminalClient, TerminalClientOptions } from './base/terminal-client';
+import { TerminalWidget, TerminalSize } from './base/terminal-widget';
+
+/**
+ * Default implementation Terminal Client.
+ */
+@injectable()
+export class DefaultTerminalClient implements TerminalClient {
+
+    @inject(ShellTerminalServerProxy)
+    protected readonly shellTerminalServer: ShellTerminalServerProxy;
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(WebSocketConnectionProvider)
+    protected readonly webSocketConnectionProvider: WebSocketConnectionProvider;
+
+    @inject(TerminalWidget)
+    readonly widget: TerminalWidget;
+
+    @inject(TerminalClientOptions)
+    readonly options: TerminalClientOptions;
+
+    @inject(TerminalWatcher)
+    protected readonly terminalWatcher: TerminalWatcher;
+
+    @inject(ILogger) @named('terminal')
+    protected readonly logger: ILogger;
+
+    private _terminalId: number = -1;
+
+    protected waitForConnection: Deferred<MessageConnection>;
+
+    protected readonly toDispose = new DisposableCollection();
+    protected readonly toDisposeOnConnect = new DisposableCollection();
+    protected onDidCloseDisposable: Disposable;
+
+    get processId(): Promise<number> {
+        return (async () => {
+            if (!IBaseTerminalServer.validateId(this.terminalId)) {
+                throw new Error('terminal is not started');
+            }
+            return this.shellTerminalServer.getProcessId(this.terminalId);
+        })();
+    }
+
+    get terminalId(): number {
+        return this._terminalId;
+    }
+
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.push(this.terminalWatcher.onTerminalError(({ terminalId, error }) => {
+            if (terminalId === this.terminalId) {
+                this.dispose();
+                this.logger.error(`The terminal process terminated. Cause: ${error}`);
+            }
+        }));
+        this.toDispose.push(this.terminalWatcher.onTerminalExit(({ terminalId }) => {
+            if (terminalId === this.terminalId) {
+                this.dispose();
+            }
+        }));
+
+       this.toDispose.push(this.widget.onTerminalResize(size => this.resize(size)));
```

Comment:
Should not this code always be in `TerminalWidget`?

Reply:
Could You elaborate more information, please? You against dispose of resize subscription? Or Do you want that terminal widget call TerminalClient#resize?

Reply:
I meant that https://github.com/theia-ide/theia/blob/7cea837f5cf9607c86cc2f65a6799e0742f4ba3b/packages/terminal/src/browser/default-terminal-client.ts#L80-L93 could stay in `TerminalWidgetImpl`.